### PR TITLE
enhancement #849 improved endpoint to get all teachers:

### LINKED ISSUE #849: Improve an Endpoint to get all teachers
Your task is to improve an Endpoint to get all teachers _(you need to add **"assigning roles"** and **"names of departments"**)_

_Buena suerte_

### PR DESCRIPTION
**Pull request by aim4ik11**
-added teacher roles
-added teacher cathedras

closes #849

### CODE REVIEW

**Review by dmykyr — APPROVED** [review]

**Review by ArtemiiLiakh — APPROVED** [review]

File: src/v2/database/entities/DbTeacher.ts [hunk 1]
```diff
@@ -8,7 +8,13 @@ export class DbTeacher {
...
   description?: string;
   avatar?: string;
-  cathedras: TeachersOnCathedras[];
+  cathedras: {
+    teacherId: string,
+    cathedraId: string,
+    cathedra: Cathedra,
+    createdAt: Date,
+    updatedAt: Date,
+  }[];
```
Comment (fokaaas) [inline src/v2/database/entities/DbTeacher.ts]:
```ts
cathedras: (TeachersOnCathedras & {
    cathedra: Cathedra,
  })[];
```

File: src/v2/api/controllers/TeacherController.ts [hunk 2]
```diff
@@ -54,16 +55,17 @@ export class TeacherController {
...
-                  Page must be a number
-                  PageSize must be a number
-                  Wrong value for order`,
+    description: `\n
+    InvalidQueryException:
+      Page must be a number
+      PageSize must be a number
+      Wrong value for order`,
   })
+  @Get()
```
Comment (fokaaas) [inline src/v2/api/controllers/TeacherController.ts:81]: Add ```@ApiEndpoint()```

**Review by fokaaas — CHANGES_REQUESTED** [review]
There are certain remarks, the endpoint does not quite match the design.
- you must have a multichoice for **cathedras**, i.e. an array of queries
- you must have a multichoice for **teacher roles**

image

**You can also extend from QueryAll**

File: src/v2/api/controllers/TeacherController.ts [hunk 3]
```diff
@@ -168,29 +183,34 @@ export class TeacherController {
...
-                  You do not have permission to perform this action`,
+    description: `\n
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiEndpoint({
+    summary: 'Create a teacher',
+    permissions: PERMISSION.TEACHERS_CREATE,
   })
   @Post()
```
Comment (fokaaas) [inline src/v2/api/controllers/TeacherController.ts:215]: if you choose to describe the documentation for this endpoint, finish it by adding descriptions to the body

File: src/v2/api/controllers/TeacherController.ts [hunk 4]
```diff
@@ -203,27 +223,35 @@ export class TeacherController {
...
-                  You do not have permission to perform this action`,
+    description: `\n
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiEndpoint({
+    summary: 'Update a teacher',
+    permissions: PERMISSION.TEACHERS_$TEACHERID_UPDATE,
   })
   @Patch('/:teacherId')
```
Comment (fokaaas) [inline src/v2/api/controllers/TeacherController.ts:256]: if you choose to describe the documentation for this endpoint, finish it by adding descriptions to the body

File: src/v2/api/pipes/AllTeachersPipe.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    private readonly cathedraByIdPipe: CathedraByIdPipe,
+  ) {}
+
+  async transform (query: QueryAllTeacherDTO): Promise<QueryAllTeacherDTO> {
+    if (query.groupId) {
+      await this.groupByIdPipe.transform(query.groupId);
+    }
+    if (query.cathedrasId) {
+      await Promise.all(query.cathedrasId?.map((cathedraId) => this.cathedraByIdPipe.transform(cathedraId)));
+    }
```
Comment (fokaaas) [inline src/v2/api/pipes/AllTeachersPipe.ts]: Transfer this logic to service or improve existing pipes

File: src/v2/mappers/TeacherMapper.ts [hunk 6]
```diff
@@ -17,21 +16,25 @@ export class TeacherMapper {
...
   getRoles (teacher: DbTeacher): Set<TeacherRole> {
-    const roles = [];
+    const roles: TeacherRole[] = [];
     for (const dt of teacher.disciplineTeachers) {
       roles.push(...dt.roles.map((r) => r.role));
     }
     return new Set(roles);
   }
-
+  getTeacherCathedras (teacher: DbTeacher) : string[] {
```
Comment (fokaaas) [inline src/v2/mappers/TeacherMapper.ts]: Make this method private and add spaces

**Review by fokaaas — CHANGES_REQUESTED** [review]

File: src/v2/api/controllers/TeacherController.ts (hunk 3 above)
Comment (aim4ik11) [reply src/v2/api/controllers/TeacherController.ts:215]: Polyak moment

**Review by fokaaas — APPROVED** [review]

**Pull request merged**